# Task 2: Data Cleaning and Preprocessing

## Objective
The objective of this task is to clean and preprocess the Titanic dataset by handling missing values, removing duplicate records, detecting and treating outliers, and normalizing numerical features. This improves data quality and prepares the dataset for analysis and machine learning applications.

## Dataset Description

The Titanic dataset contains information about passengers aboard the RMS Titanic. The dataset includes demographic details, ticket information, fare, passenger class, and survival status.

### Features:
- PassengerId
- Survived
- Pclass
- Name
- Sex
- Age
- SibSp
- Parch
- Ticket
- Fare
- Cabin
- Embarked

## Import Required Libraries

The necessary Python libraries are imported for data manipulation, preprocessing, and normalization.

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

## Load Dataset

Load the Titanic dataset into a Pandas DataFrame for further analysis and preprocessing.

In [2]:
df = pd.read_csv("Titanic-Dataset.csv")
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Dataset Shape

Display the number of rows and columns present in the dataset.

In [3]:
df.shape

(891, 12)

## Dataset Information

Examine the dataset structure, data types, non-null values, and memory usage.

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


## Statistical Summary

Generate descriptive statistics to understand the distribution, central tendency, and spread of numerical features in the dataset.

In [5]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## Check Missing Values

Identify missing values present in each column of the dataset.

In [6]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Handle Missing Values

Missing values are treated using appropriate techniques:
- Age is filled using the median value.
- Embarked is filled using the mode value.
- Cabin is removed because it contains a large number of missing values.

In [7]:
# Fill missing values in Age using median
df["Age"] = df["Age"].fillna(df["Age"].median())

# Fill missing values in Embarked using mode
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

# Drop Cabin column due to excessive missing values
df.drop("Cabin", axis=1, inplace=True)

print("Missing values handled successfully")

Missing values handled successfully


## Verify Missing Values

Confirm that all missing values have been handled successfully.

In [8]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

## Check Duplicate Records

Identify duplicate rows present in the dataset.

In [9]:
df.duplicated().sum()

np.int64(0)

## Detect Outliers

The Interquartile Range (IQR) method is used to identify outliers in the Fare column.

In [10]:
Q1 = df["Fare"].quantile(0.25)
Q3 = df["Fare"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Lower Bound: -26.724
Upper Bound: 65.6344


## Handle Outliers

Records containing Fare values outside the acceptable IQR range are removed.

In [11]:
original_shape = df.shape

df = df[(df["Fare"] >= lower_bound) & (df["Fare"] <= upper_bound)]

print("Original Shape:", original_shape)
print("New Shape:", df.shape)

Original Shape: (891, 11)
New Shape: (775, 11)


## Normalize Features

Numerical features are scaled using MinMaxScaler to bring values into a common range between 0 and 1.

In [12]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df[["Age", "Fare"]] = scaler.fit_transform(df[["Age", "Fare"]])

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,0.271174,1,0,A/5 21171,0.111538,S
2,3,1,3,"Heikkinen, Miss. Laina",female,0.321438,0,0,STON/O2. 3101282,0.121923,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,0.434531,1,0,113803,0.816923,S
4,5,0,3,"Allen, Mr. William Henry",male,0.434531,0,0,373450,0.123846,S
5,6,0,3,"Moran, Mr. James",male,0.346569,0,0,330877,0.130128,Q


## Save Clean Dataset

The cleaned and preprocessed dataset is saved as a CSV file for future analysis.

In [13]:
df.to_csv("cleaned_titanic.csv", index=False)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


## Final Dataset Preview

Display the first few rows of the cleaned dataset.

In [14]:
print("Final Shape:", df.shape)
df.head()

Final Shape: (775, 11)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,0.271174,1,0,A/5 21171,0.111538,S
2,3,1,3,"Heikkinen, Miss. Laina",female,0.321438,0,0,STON/O2. 3101282,0.121923,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,0.434531,1,0,113803,0.816923,S
4,5,0,3,"Allen, Mr. William Henry",male,0.434531,0,0,373450,0.123846,S
5,6,0,3,"Moran, Mr. James",male,0.346569,0,0,330877,0.130128,Q


## Observations

- The original dataset contained 891 rows and 12 columns.
- Missing values were found in Age, Cabin, and Embarked.
- Age values were filled using the median.
- Embarked values were filled using the mode.
- The Cabin column was removed due to a high percentage of missing values.
- Outliers in the Fare column were detected and removed using the IQR method.
- Numerical features were normalized using MinMaxScaler.
- The final cleaned dataset contains 775 rows and 11 columns.

## Conclusion

The Titanic dataset was successfully cleaned and preprocessed. Missing values were handled, duplicate records were checked and removed if present, outliers were treated using the IQR method, and numerical features were normalized. The cleaned dataset is now ready for further analysis and machine learning applications.